compiler/pipes/calc-bad-vars.cpp

```diff
@@ -115,13 +115,297 @@ struct FuncCallGraph {
+    auto actual_it = actual_chain.end() - 1;
+
+    while (true) {
+      bool rightmost_matched = *rule_it == *actual_it;
+
+      if (rightmost_matched) {
+        if (rule_it == rule_chain.begin()) {
+          return true;
+        }
+        if (actual_it == actual_chain.begin()) {
+          return false;
+        }
+        --rule_it;
+        --actual_it;
+      } else {
+        if (actual_it == actual_chain.begin()) {
+          return false;
+        }
+        --actual_it;
+      }
+    }
+  }
+
+  // on error (colored chain breaks some rule), we want to find an actual chain of calling
+  // this is launched only on error, that's why we don't care about performance and just use bfs
+  // todo think about using such approach for fork chains, throws chains, etc (and drop fork_prev, wait_prev, throws_reason from FunctionData)
+  std::vector<FunctionPtr> find_callstack_between_two_functions_bfs(FunctionPtr from, FunctionPtr target, const std::unordered_set<FunctionPtr> &shouldnt_appear) {
+    std::unordered_map<FunctionPtr, int> visited_level;
+    std::queue<FunctionPtr> bfs_queue;
+
+    bfs_queue.push(from);
+    visited_level[from] = 0;
+
+    while (!bfs_queue.empty()) {
+      FunctionPtr cur = bfs_queue.front();
+      bfs_queue.pop();
+      int next_level = visited_level[cur] + 1;
+
+      if (cur == target) {
+        break;
+      }
+
+      for (FunctionPtr called : call_graph.graph[cur]) {
+        if (shouldnt_appear.find(called) == shouldnt_appear.end() && visited_level.emplace(called, next_level).second) {
+          bfs_queue.push(called);
+        }
+      }
+    }
+
+    if (visited_level.find(target) == visited_level.end()) {    // couldn't find, just return [from, target]
+      return {from, target};
+    }
+
+    std::vector<FunctionPtr> callstack;
+    callstack.emplace_back(target);
+    for (FunctionPtr cur = target; cur != from;) {
+      int prev_level = visited_level[cur] - 1;
+      for (FunctionPtr callee : call_graph.rev_graph[cur]) {
+        auto found = visited_level.find(callee);
+        if (found != visited_level.end() && found->second == prev_level) {
+          callstack.push_back(callee);
+          cur = callee;
+          break;
+        }
+      }
+      kphp_assert(visited_level.find(cur)->second == prev_level);
+    }
+
+    std::reverse(callstack.begin(), callstack.end());
+    return callstack;
+  }
+
+  void show_error_on_rule_broken(const CallstackOfColoredFunctions &color_callstack, const function_palette::PaletteRule &rule) {
+    kphp_assert(rule.is_error());
+    if (shown_errors.size() > max_shown_errors_count) {
+      return;
+    }
+
+    auto vector = color_callstack.as_vector();    // f1, f2, f3 — all of them are colored, and their chain breaks the rule
+    std::vector<FunctionPtr> full_callstack;      // will be: src_main -> ... -> f1 -> ... -> f2 -> ... -> f3
+    for (int i = 0; i < vector.size() - 1; ++i) {
+      FunctionPtr cur = vector[i], next = vector[i+1];
```

Comment:
If vector is empty, `vector.size() - 1` is `2^64 - 1` because of unsigned type. So I would prefer rewrite it like this:
```suggestion
    for (int i = 1; i < vector.size(); ++i) {
      FunctionPtr cur = vector[i - 1], next = vector[i];
```

Reply:
Maybe cast to int is easier here?
We have similar explicitly cast pieces of code.

Reply:
I think, the option without `- 1` in size calculation will look more clear and graceful. Also I can't say that it makes code more complicated somehow. Why do you think cast is easier?

Reply:
Loop
```cpp
for (int i = 1; i < vector.size(); ++i) {
```
does not look like a loop until the penultimate element, to understand this you need to look at the code further.

In a loop with an explicit cast, you can see it right away.
```cpp
for (int i = 0; i < static_cast<int>(vector.size()) - 1; ++i) {
```

Reply:
Ok. Alternatively you can just add `assert(!vector.empty())` before the loop and don't add explicit cast

Reply:
Replace with assert.